### PR TITLE
A command line tool to do various Leela related stuff.

### PR DESCRIPTION
**Pull request by mooskagh**
Supports `leela2onnx`, `onnx2leela` and `describenet` commands.

Example run:

```
~/d/l/b/debugoptimized (ctl↑2|✚5…) $ ./lc0ctl leela2onnx --input=$HOME/my/Jupyter/lczero/vis/nobackup/id69291 --output=moo
Lc0 tool v0.29.0-dev+git.dirty built Oct 30 2021

Weights
~~~~~~~
                      Blocks: 30
                     Filters: 384
                      Policy: Convolution
                       Value: WDL
                         MLH: Present
Converting Leela network to the ONNX.

ONNX interface
~~~~~~~~~~~~~~
                Input planes: /input/planes
                  Output WDL: /output/wdl
               Output Policy: /output/policy
                  Output MLH: /output/mlh

Format
~~~~~~
                       Input: INPUT_CLASSICAL_112_PLANE
                     Network: NETWORK_ONNX
                      Policy: POLICY_CLASSICAL
                       Value: VALUE_WDL
                         MLH: MOVES_LEFT_V1
Done.
~/d/l/b/debugoptimized (ctl↑2|✚5…) $ ./lc0ctl onnx2leela --input=moo --output=foo --onnx-input=/input/planes --onnx-output-policy=/output/policy --onnx-output-wdl=/output/wdl
Lc0 tool v0.29.0-dev+git.dirty built Oct 30 2021
Warning: ONNX output node '/output/mlh' not used.

Format
~~~~~~
                       Input: INPUT_CLASSICAL_112_PLANE
                     Network: NETWORK_ONNX
                      Policy: POLICY_CLASSICAL
                       Value: VALUE_WDL

ONNX interface
~~~~~~~~~~~~~~
                   Data type: FLOAT
                Input planes: /input/planes
                  Output WDL: /output/wdl
               Output Policy: /output/policy

ONNX model
~~~~~~~~~~
                  IR version: 4
               Producer Name: Lc0
            Producer Version: 0.29.0-dev+git.dirty
                      Domain: org.lczero.models.*
                       Input: /input/planes
                      Output: /output/policy
                      Output: /output/wdl
                      Output: /output/mlh
                       Opset: 9
Done.
~/d/l/b/debugoptimized (ctl↑2|✚5…) $ ./lc0ctl describenet --weights=foo
Lc0 tool v0.29.0-dev+git.dirty built Oct 30 2021

General
~~~~~~~
         Minimal Lc0 version: v0.28.0

Format
~~~~~~
                       Input: INPUT_CLASSICAL_112_PLANE
                     Network: NETWORK_ONNX
                      Policy: POLICY_CLASSICAL
                       Value: VALUE_WDL

ONNX interface
~~~~~~~~~~~~~~
                   Data type: FLOAT
                Input planes: /input/planes
                  Output WDL: /output/wdl
               Output Policy: /output/policy

ONNX model
~~~~~~~~~~
                  IR version: 4
               Producer Name: Lc0
            Producer Version: 0.29.0-dev+git.dirty
                      Domain: org.lczero.models.*
                       Input: /input/planes
                      Output: /output/policy
                      Output: /output/wdl
                      Output: /output/mlh
                       Opset: 9
```

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -699,3 +700,20 @@ if get_option('python_bindings')
...
+
+if get_option('lc0ctl')
+  lc0ctl_files = [
+    'src/lc0ctl/describenet.cc',
+    'src/lc0ctl/leela2onnx.cc',
+    'src/lc0ctl/onnx2leela.cc',
+  ]
+
+  executable('lc0ctl', 'src/lc0ctl/lc0ctl_main.cc',
+    files + lc0ctl_files,
```
Comment (borg323) [inline meson.build:716]: This is a bit wasteful. What about something like:
```diff
diff --git a/meson.build b/meson.build
index 6a9853c..25f0c57 100644
--- a/meson.build
+++ b/meson.build
@@ -633,8 +633,10 @@ if get_option('embed')
   add_project_arguments('-DEMBED', language : 'cpp')
 endif
 
+lc0_lib = static_library('lc0_lib', files, include_directories: includes, dependencies: deps)
+
 executable('lc0', 'src/main.cc',
-  files, include_directories: includes, dependencies: deps, install: true)
+  include_directories: includes, link_with: lc0_lib, install: true)
 
 #############################################################################
 ## Tests
@@ -642,7 +644,6 @@ executable('lc0', 'src/main.cc',
 
 if get_option('gtest')
   gtest = dependency('gtest', fallback: ['gtest', 'gtest_dep'])
-  lc0_lib = library('lc0_lib', files, include_directories: includes, dependencies: deps)
 
   test('ChessBoard',
     executable('chessboard_test', 'src/chess/board_test.cc',
@@ -713,7 +714,7 @@ if get_option('lc0ctl')
   ]
 
   executable('lc0ctl', 'src/lc0ctl/lc0ctl_main.cc',
-    files + lc0ctl_files,
-    include_directories: includes, dependencies: deps, install: true)
+    lc0ctl_files,
+    include_directories: includes, link_with: lc0_lib, install: true)
 
 endif
```

**Review by borg323 — COMMENTED** [review]
A few first comments.

File: meson.build (hunk 1 above)
Comment (borg323) [reply meson.build:716]: This seems to work even without an older `net.pb.h` but I don't like it:
```diff
diff --git a/meson.build b/meson.build
index 6a9853c..6961e24 100644
--- a/meson.build
+++ b/meson.build
@@ -90,12 +90,11 @@ if run_command('scripts/checkdir.py', 'libs/lczero-common/proto').returncode() !
   endif
 endif
 
+files += 'src/utils/protomessage.cc'
 pb_files = [
-  'src/utils/protomessage.cc',
   gen.process('libs/lczero-common/proto/net.proto',
     preserve_path_from : meson.current_source_dir() + '/libs/lczero-common/')
 ]
-files += pb_files
 
 # Extract git short revision.
 short_rev = 'unknown'
@@ -146,7 +145,7 @@ gen_proto_src = generator(compile_proto, output: ['@BASENAME@.pb.h'],
     '--cpp_out=@BUILD_DIR@',
     '@INPUT@'])
 
-files += gen_proto_src.process('src/neural/onnx/onnx.proto',
+pb_files += gen_proto_src.process('src/neural/onnx/onnx.proto',
   preserve_path_from : meson.current_source_dir() + '/src/')
 
 #############################################################################
@@ -633,8 +632,10 @@ if get_option('embed')
   add_project_arguments('-DEMBED', language : 'cpp')
 endif
 
+lc0_lib = static_library('lc0_lib', pb_files, files, include_directories: includes, dependencies: deps)
+
 executable('lc0', 'src/main.cc',
-  files, include_directories: includes, dependencies: deps, install: true)
+  pb_files, include_directories: includes, link_with: lc0_lib, install: true)
 
 #############################################################################
 ## Tests
@@ -642,7 +643,6 @@ executable('lc0', 'src/main.cc',
 
 if get_option('gtest')
   gtest = dependency('gtest', fallback: ['gtest', 'gtest_dep'])
-  lc0_lib = library('lc0_lib', files, include_directories: includes, dependencies: deps)
 
   test('ChessBoard',
     executable('chessboard_test', 'src/chess/board_test.cc',
@@ -713,7 +713,7 @@ if get_option('lc0ctl')
   ]
 
   executable('lc0ctl', 'src/lc0ctl/lc0ctl_main.cc',
-    files + lc0ctl_files,
-    include_directories: includes, dependencies: deps, install: true)
+    pb_files, lc0ctl_files,
+    include_directories: includes, link_with: lc0_lib, install: true)
 
 endif
```
An alternative is:
```diff
diff --git a/meson.build b/meson.build
index 6a9853c..7b3b51f 100644
--- a/meson.build
+++ b/meson.build
@@ -66,11 +66,6 @@ includes += include_directories('third_party', is_system: true)
 
 # Compiling protobufs.
 compile_proto = find_program('scripts/compile_proto.py')
-gen = generator(compile_proto, output: ['@BASENAME@.pb.h'],
-  arguments : [
-    '--proto_path=@CURRENT_SOURCE_DIR@/libs/lczero-common',
-    '--cpp_out=@BUILD_DIR@',
-    '@INPUT@'])
 
 # Handle submodules.
 git = find_program('git', required: false)
@@ -90,12 +85,16 @@ if run_command('scripts/checkdir.py', 'libs/lczero-common/proto').returncode() !
   endif
 endif
 
+files += 'src/utils/protomessage.cc'
 pb_files = [
-  'src/utils/protomessage.cc',
-  gen.process('libs/lczero-common/proto/net.proto',
-    preserve_path_from : meson.current_source_dir() + '/libs/lczero-common/')
+    custom_target(
+    'net.pb.h',
+    output : 'net.pb.h',
+    input : 'libs/lczero-common/proto/net.proto',
+    command : [compile_proto, '--proto_path=@CURRENT_SOURCE_DIR@/libs/lczero-common',
+        '--cpp_out=@BUILD_ROOT@', '@INPUT@']
+    )
 ]
-files += pb_files
 
 # Extract git short revision.
 short_rev = 'unknown'
@@ -140,14 +139,13 @@ elif get_option('malloc') != ''
 endif
 
 # ONNX protobufs.
-gen_proto_src = generator(compile_proto, output: ['@BASENAME@.pb.h'],
-  arguments : [
-    '--proto_path=@CURRENT_SOURCE_DIR@/src',
-    '--cpp_out=@BUILD_DIR@',
-    '@INPUT@'])
-
-files += gen_proto_src.process('src/neural/onnx/onnx.proto',
-  preserve_path_from : meson.current_source_dir() + '/src/')
+pb_files += custom_target(
+    'onnx.pb.h',
+    output : 'onnx.pb.h',
+    input : 'src/neural/onnx/onnx.proto',
+    command : [compile_proto, '--proto_path=@CURRENT_SOURCE_DIR@/src',
+        '--cpp_out=@BUILD_ROOT@', '@INPUT@'],
+)
 
 #############################################################################
 ## Main files
@@ -633,8 +631,10 @@ if get_option('embed')
   add_project_arguments('-DEMBED', language : 'cpp')
 endif
 
+lc0_lib = static_library('lc0_lib', pb_files, files, include_directories: includes, dependencies: deps)
+
 executable('lc0', 'src/main.cc',
-  files, include_directories: includes, dependencies: deps, install: true)
+  pb_files, include_directories: includes, link_with: lc0_lib, install: true)
 
 #############################################################################
 ## Tests
@@ -642,7 +642,6 @@ executable('lc0', 'src/main.cc',
 
 if get_option('gtest')
   gtest = dependency('gtest', fallback: ['gtest', 'gtest_dep'])
-  lc0_lib = library('lc0_lib', files, include_directories: includes, dependencies: deps)
 
   test('ChessBoard',
     executable('chessboard_test', 'src/chess/board_test.cc',
@@ -713,7 +712,7 @@ if get_option('lc0ctl')
   ]
 
   executable('lc0ctl', 'src/lc0ctl/lc0ctl_main.cc',
-    files + lc0ctl_files,
-    include_directories: includes, dependencies: deps, install: true)
+    pb_files, lc0ctl_files,
+    include_directories: includes, link_with: lc0_lib, install: true)
 
 endif
```
Probably too much trouble, just moving all the tools into lc0 is starting to look tempting.

Comment (borg323) [reply meson.build:716]: Using a library like that messes the backend registration (which in retrospect should have been obvious). I had done this right in #1031...

File: src/utils/files.cc [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+#include <zlib.h>
+
+#include <cstdio>
+
+#include "utils/exception.h"
+
+namespace lczero {
+
+std::string ReadFileToString(const std::string& filename) {
```
[the remark this answers is not in the document]
Comment (borg323) [reply src/utils/files.cc:36]: Did a quick mock-up and it seems to work just fine.

File: meson.build (hunk 1 above)
Comment (mooskagh) [reply meson.build:716]: Just merged everything into a single binary as discussed.

File: src/utils/files.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (mooskagh) [reply src/utils/files.cc:36]: Done.

File: src/lc0ctl/onnx2leela.cc [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+  If you modify this Program, or any covered work, by linking or
+  combining it with NVIDIA Corporation's libraries from the NVIDIA CUDA
+  Toolkit and the NVIDIA CUDA Deep Neural Network library (or a
+  modified version of those libraries), containing parts covered by the
+  terms of the respective license agreement, the licensors of this
+  Program grant you additional permission to convey the resulting work.
+*/
+
+#include <zlib.h>
+
```
[the remark this answers is not in the document]
Comment (mooskagh) [reply src/lc0ctl/onnx2leela.cc:29]: Done.

File: src/utils/files.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (mooskagh) [reply src/utils/files.cc:36]: Done.

File: meson_options.txt [hunk 4]
```diff
@@ -172,3 +172,8 @@ option('onnx_include',
        type: 'string',
        value: '',
        description: 'Paths to ONNX runtime includes')
+
+option('lc0ctl',
```
Comment (borg323) [inline meson_options.txt:176]: This isn't needed any more.

**Review by borg323 — APPROVED** [review]
The ` meson_options.txt` change is not needed any more.

File: meson_options.txt (hunk 4 above)
Comment (mooskagh) [reply meson_options.txt]: Done.

**Pull request merged**